### ISSUE TITLE
kstatus: defaults to fail open when no `status` field is available

### ISSUE THREAD

**Issue by jack1902** [kubernetes-sigs/cli-utils#632]
TL;DR - When a resource doesn't yet have a `status` field does kstatus report `unknown` ?

When creating a resource which is defined by a CRD which has no `status` by default, i believe the behaviour of kstatus is to state "unknown". I believe this then enables developers using `unknown` as a condition to determine if that should be seen as "lets carry on" or "stop here and wait".

If the "unknown" status is indeed what happens when no `status` field exists on a resource (because it has been freshly applied and the backing controller hasn't had a chance to update the CR' status field yet, i'll close this issue and raise another with the project using this one to add configuration.

**Comment by Raffo** [comment]
One thing that we have been doing is to patch the logic for every custom CRD that we create. It's not great, but we also don't have a lot of custom CRDs.

**Comment by aw185176** [comment]
> When a resource doesn't yet have a status field does kstatus report unknown ?

If I am reading the code right (https://github.com/kubernetes-sigs/cli-utils/blob/0b156cb0425fdb29a436d13f840a39039558c10e/pkg/kstatus/status/status.go#L102), it actually falls all the way through to `Current` if there is no status field.

**Comment by karlkfi** [comment]
What's your use case?

AFAICT, the kstatus when there's no status field populated is "Current". 

However, if you have an optional non-pointer status field configured in the CRD, then the server sets status to an empty map by default, which will cause kstatus to interrupt the status as "InProgress" UNLESS status.observedGeneration == metadata.generation AND there aren't any status.conditions that get interpreted differently. 

So the recommended pattern is to add status.observedGeneration to the CRD with a default value of 0.

Unfortunately, all other proposed solution I've heard of require custom kstatus code or custom kstatus providers (NYI).

But if you have a specific if. Use case, let us know.

**Comment by rohitsakala** [comment]
I have the same use case as this issue. Nothing special. 

Just curious to understand why can't we check if the resource has a status sub resource or not ? and if does have a status sub resource, we return it as `In Progress`. 

Say for example when I do `kubectl get --raw /apis/<group>/<version> | jq`

I get the following response 

```
 {
      "name": "kind/status",
      "singularName": "",
      "namespaced": false,
      "kind": "kind",
      "verbs": [
        ....
      ]
  },
```

This tells us that this resource has status subresource. 

Thank you for your reply. @karlkfi

**Comment by karlkfi** [comment]
There's a couple problems with that strategy, but the big one is that not all resources are custom resources and the resource discovery API does not expose this information. So you have to know that it's a custom resource and then listwatch & cache all the CRDs in memory to speed up the lookup you need to know whether there is or isn't a status field in the API. 

As a bonus problem, just because it's not a CRD doesn't mean it's a built-in resource. There are lots of APIServices that expose custom resources without an equivalent CRD API to lookup the resource schema.

We can hard-code expectations for built-in resources, but for everything else we really need a way to configure custom rules. 

That said, it would hypothetically be possible to solve this for CRDs, but it would add API calls and increase memory usage. kpt-style CLI commands would need to get the CRDs on every command execution or cache them locally on disk. Continuous gitops services like Config Sync would have to increase API calls and memory requirements to watch and cache every CRD in memory with or without another layer of disk caching.
